### ISSUE TITLE
RangeError - Maximum call stack size exceeded.

### ISSUE THREAD

**Issue by alexanderwallin** [gorhom/react-native-portal#15]
Hi there,

I liked the API design of this library and decided to use it together with [`react-native-modalize`](https://github.com/jeremybarbet/react-native-modalize). It worked well locally, but in distribution builds the below error caused crashes on both iOS and Android.

I switched to another library and unfortunately can't put time into creating a bare minimum reproducible case. But I thought I'd share it!

```
RangeError
Maximum call stack size exceeded.

    node_modules/immer/dist/immer.esm.js:1:712 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588 
    node_modules/immer/dist/immer.esm.js:1:782 
    [native code] forEach
    node_modules/immer/dist/immer.esm.js:1:745 i
    node_modules/immer/dist/immer.esm.js:1:2562 M
    node_modules/immer/dist/immer.esm.js:1:3055 A
    node_modules/immer/dist/immer.esm.js:1:2588
```

**Comment by gorhom** [comment]
hi @alexanderwallin, could you provide a reproducible sample code so i could investigate it.

**Comment by alexanderwallin** [comment]
@gorhom Sorry, like I said in the description, I unfortunately don't have the time.

**Comment by beqramo** [comment]
@gorhom  I have the same problem, and reproduction steps are simple, take any example and place Portalhost in the parent component of  Portal and you will see the problem. I agree @alexanderwallin  that it is only in production.

**Comment by gorhom** [comment]
@beqramo thanks for adding more context to this bug,  i will look into it

**Comment by beqramo** [comment]
@gorhom by the way, I'm using it with https://github.com/dabakovich/react-native-controlled-mentions#readme
maybe it will give you some clue, and maybe the problem is when using it with a library.

**Comment by beqramo** [comment]
indeed I was able to resolve this issue by wrapping my components in useMemo. there was rerender problem and so for me, that was a solution.

**Comment by mobily** [comment]
@alexanderwallin I've just tested it, and it still doesn't work, the same error occurs 😢

@beqramo cool! nevertheless, I think it should work either with or without `memo`

**Comment by mobily** [comment]
@gorhom yup, I have been looking for the root of this issue for a few hours and couldn't find anything, and removing `Immer` did solve the issue eventually

**Comment by gorhom** [comment]
@mobily alright! one more thing, could you provide a reproducible sample code ?

**Comment by mobily** [comment]
@gorhom sure, I will create a repo later today ;)

**Comment by timbastin** [comment]
I am having the same issue. Is there any fix for this yet?

**Comment by gorhom** [comment]
@mobily any luck with the repo ?

**Comment by mobily** [comment]
@gorhom 👋 no luck :( I have created a fresh RN project with basic components/hooks/helpers that I use in my commercial project, and it works with no error, the app doesn't crash in the release mode… not sure why, I'm looking for a reason

**Comment by jordoh** [comment]
I'm seeing the same exception using react-native-portal via react-native-bottom-sheet - it only happens in production, when a button on one `BottomSheetModal` triggers a `present` on a second `BottomSheetModal`. I've spent some time trying to reproduce in expo (with expo in [production mode](https://docs.expo.io/workflow/development-mode/#production-mode)) without any success.

**Comment by alexco2** [comment]
I have the same error in react-native-bottom-sheet. It happens when I close the screen that holds the `bottomSheetModalRef` while the bottom-sheet modal is still open (for example with the hardware back button). In that case the BottomSheetModal closes after the screen is closed. If I try to open the BottomSheetModal again after opening the screen, the app crashes with the error "RangeError: Maximum call stack size exceeded" error.
I tried to close the BottomSheetModal activley when the screen unmounts with useeffect, which did not solve the problem. If I listen for the Backhandler and close the BottomSheetModal as soon as the BackButton is pressed, the error just occurs, if I fully extended the BottomSheetModal first. Then the BottomSheetModal is closed when pressing the BackButton, but can not be opened again. After closing and opening the screen that holds the `bottomSheetModalRef`, the app crashes when I try to open the BottomSheetModal again.
I hope this description somehow helps.

**Comment by gorhom** [comment]
Hi everyone , could anyone try to reproduce it on `v1.0.6`, thanks ?

**Comment by egadstar** [comment]
I had seen this error before, but no more. Thanks.

**Comment by mobily** [comment]
@gorhom works fine! thanks for fixing this issue ❤️

**Comment by gorhom** [comment]
@mobily thanks for testing it 👏

**Comment by dburdan** [comment]
We're also getting `Maximum call stack size exceeded` when dismissing one `BottomSheetModal` and immediately opening another. Using `@gorhom/portal: 1.0.7` & `@gorhom/bottom-sheet": 4.0.0-alpha.18`. Only happens in release mode, and haven't been able to find a solution.

**Comment by gorhom** [comment]
@dburdan could you please provide more info ( env, os, rn version ) or a reproducible sample ? I appreciate it

**Comment by dburdan** [comment]
@gorhom We are running `react-native: 0.64.2`. Crash exists across iOS and Android, and every device we've tested.

It's been difficult to trace since everything works great when in debug mode. The crash only appears after building to device or to simulator in release mode. Is there any other information I could share that would be helpful? If I get some time this week, I'll try to build a reproducible example.

**Comment by UrbanChrisy** [comment]
I was using react `context provider` inside the `Portal` component, and I was getting this issue in a release build only like above. When I moved the `context provider` to the outside of the `PortalHost`. Because I'm using the portal host at a different layer than the root host. Then I no longer had this issue. I don't know what would cause this, but ill try to create a reproducible expo demo tomorrow. Ill update where I get to.

**Comment by UrbanChrisy** [comment]
Unable to reproduce as of yet. Ive spent wayyy too long on this already sorry guys unable to give a demo.

**Comment by lucianojsjr** [comment]
@gorhom Are there some news about this problem? I'm getting the same error and I'm using `react-native@0.64.1`.

**Comment by SCasarotto** [comment]
I am also seeing this issue but it only seems to appear once the app is in release mode or built.

**Comment by egadstar** [comment]
Could this be the source based on the original stacktrace? https://github.com/immerjs/immer/issues/832

**Comment by egadstar** [comment]
I ended up forking and refactoring immer from the project; my production builds are no longer crashing with the RangeError, so feel pretty certain that the issue is related to a bug within immer.

**Comment by gorhom** [comment]
@egadstar would you mind submitting a pr :)

**Comment by egadstar** [comment]
Yea, no problem. Because I removed immer, I didn't know if you wanted to move that direction. I'll get it submitted, and you can review it.

**Comment by egadstar** [comment]
I did a poor job reading through this thread; didn't realize that @mobily had already zeroed in on immer.

**Comment by gorhom** [comment]
this should be fixed with `v1.0.9` , thanks to @egadstar PR 👏
